File: 238. Product of Array Except Self/Two-Pass-Prefix-Suffix.py
def productExceptSelf(nums):
    n = len(nums)
    ans = [0] * n
    
    # First pass: Compute prefix products
    left = 1
    for i in range(n):
        ans[i] = left
        left *= nums[i]
    
    # Second pass: Compute suffix products and multiply
    right = 1
    for i in range(n-1, -1, -1):
        ans[i] *= right
        right *= nums[i]
    
    return ans

File: 238. Product of Array Except Self/Two-Pass-Prefix-Suffix.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];
        
        // First pass: Compute prefix products (left products)
        int left = 1;
        for (int i = 0; i < n; i++) {
            ans[i] = left;
            left *= nums[i];
        }
        
        // Second pass: Compute suffix products (right products) and multiply
        int right = 1;
        for (int i = n - 1; i >= 0; i--) {
            ans[i] *= right;
            right *= nums[i];
        }
        
        return ans;
    }
}